keras_cv/losses/dice.py

```diff
@@ -0,0 +1,592 @@
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+
+from keras_cv.utils import losses_utils
+
+
+def _calculate_dice_numerator_denominator(y_true, y_pred, beta, axis, epsilon):
+    # Calculate Dice Loss' numerator and denominator from equation.
+    true_positive = tf.reduce_sum(y_true * y_pred, axis=axis)
+    false_positive = tf.reduce_sum(y_pred, axis=axis) - true_positive
+    false_negative = tf.reduce_sum(y_true, axis=axis) - true_positive
+
+    power_beta = 1 + beta**2
+    numerator = power_beta * true_positive + epsilon
+    denominator = (
+        (power_beta * true_positive)
+        + (beta**2 * false_negative)
+        + false_positive
+        + epsilon
+    )
+    return (numerator, denominator)
+
+
+def _generalized_dice_score(y_true, numerator, denominator, axis):
+    # Calculate the volume of ground truth labels.
+    weight = tf.math.reciprocal(tf.square(tf.reduce_sum(y_true, axis=axis)))
+
+    # Calculate the weighted dice score and normalizer.
+    weighted_numerator = tf.reduce_sum(weight * numerator)
+    weighted_denominator = tf.reduce_sum(weight * denominator)
+    general_dice_score = weighted_numerator / weighted_denominator
+
+    return general_dice_score
+
+
+def _adaptive_dice_score(numerator, denominator):
+    # Calculate the dice scores
+    dice_score = numerator / denominator
+    # Calculate weights based on Dice scores.
+    weights = tf.exp(-1.0 * dice_score)
+    # Multiply weights by corresponding scores and get sum.
+    weighted_dice = tf.reduce_sum(weights * dice_score)
+    # Calculate normalization factor.
+    normalizer = tf.cast(tf.size(input=dice_score), dtype=tf.float32) * tf.exp(-1.0)
+    # normalize the dice score
+    norm_dice_score = weighted_dice / normalizer
+
+    return norm_dice_score
+
+
+def _check_input_params(beta, loss_type=None, class_ids=None):
+    if beta <= 0.0:
+        raise ValueError(f"`beta` value should be greater than zero. Got {beta}")
+
+    if loss_type is not None:
+        if loss_type not in ["generalized", "adaptive"]:
+            raise ValueError(
+                "The `loss_type` is not valid. "
+                "If `loss_type` is not `None`, It should be either "
+                f"`generalized` or `adaptive`. Got {loss_type}"
+            )
+
+    if class_ids is not None:
+        if isinstance(class_ids, float) or any(isinstance(x, float) for x in class_ids):
+            raise ValueError(
+                f"The indices should be int or a list of integer. Got {class_ids}"
+            )
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class CategoricalDice(tf.keras.losses.Loss):
```

Comment:
A high-level recommendation -- can you split this up into separate PRs? For example, if you only use SparseDice, maybe start with that first. This greatly help us to expedite the review process (given we want to review for correctness as well)

Reply:
I hope that we could enforce this policy also with internal sourcing PRs.

Recently I saw very large PRs.

Reply:
Sure! They were in one originally, so I kept them that way. Separating into three PRs.
Do I also do Dice as a metric too?

Reply:
Also, one file for all three PRs or three files? It might be a bit verbose for the FS if there are three files given the shared method(s) between all three losses

Reply:
Separated into three PRs, one for each loss type. Since they share two methods, I've put them all in the same file. We could have a separate file for util methods, but no other losses have this format.
Is this okay? @tanzhenyu

Reply:
Chaining onto @bhack's comment - what's a large PR?
I'll make mine smaller going forward, but some wrapper would help in trying to make them more manageable for reviews :)

Reply:
Can we revisit this since it should be near the finish line? @tanzhenyu